migrations/2019-02-18T214320Z_init/up.sql

```diff
@@ -46,11 +47,31 @@ CREATE CONSTRAINT TRIGGER resource_has_parent_check
     DEFERRABLE INITIALLY DEFERRED
     FOR EACH ROW EXECUTE PROCEDURE resource_has_parent();
 
+-- Define a function to return some random bytea with the given length.
+CREATE OR REPLACE FUNCTION random_bytea(bytea_length integer) RETURNS bytea LANGUAGE sql AS
+$$
+    SELECT decode(string_agg(lpad(to_hex(width_bucket(random(), 0, 1, 256)-1),2,'0') ,''), 'hex')
+    FROM generate_series(1, $1);
+$$;
+
 -- Define a trigger function which fills in the resource name from the path.
 CREATE OR REPLACE FUNCTION resource_path() RETURNS TRIGGER LANGUAGE plpgsql AS
 $$
+DECLARE found integer;
 BEGIN
     NEW.name = (ltree2text(subpath(NEW.path, -1)));
+
+    LOOP
+        NEW.tag = encode(random_bytea(6), 'base64');
+        NEW.tag = replace(NEW.tag, '/', '_');
+        NEW.tag = replace(NEW.tag, '+', '-');
+        -- guarantee uniqueness
```

Comment:
Just to note that a `SELECT` under `READ COMMITTED` isolation level cannot avoid unique constraint check errors caused by concurrent transactions, just the probability here is very very very low. And the purpose is to generate a new one if it conflicts with an existing one, so this is okay.

Reply:
Ah good point. Yeah at least those cases will be sufficiently rare; virtually always this should run once and be done, and at least limits collisions to concurrent updates, where the probability should be sufficiently low.